homeassistant/components/broadlink/device.py

```diff
@@ -0,0 +1,57 @@
+"""Support for Broadlink devices."""
+from functools import partial
+import logging
+
+from broadlink.exceptions import (
+    AuthorizationError,
+    BroadlinkException,
+    ConnectionClosedError,
+    DeviceOfflineError,
+)
+
+from .const import DEFAULT_RETRY
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class BroadlinkDevice:
+    """Manages a Broadlink device."""
+
+    def __init__(self, hass, api):
+        """Initialize the device."""
+        self.hass = hass
+        self.api = api
+        self.available = None
+
+    async def async_connect(self):
+        """Connect to the device."""
+        try:
+            await self.hass.async_add_executor_job(self.api.auth)
+        except BroadlinkException as err_msg:
+            if self.available:
+                self.available = False
+                _LOGGER.warning(
+                    "Disconnected from device at %s: %s", self.api.host[0], err_msg
+                )
+            return False
+        else:
+            if not self.available:
+                if self.available is not None:
+                    _LOGGER.warning("Connected to device at %s", self.api.host[0])
+                self.available = True
+            return True
+
+    async def async_request(self, function, *args, **kwargs):
+        """Send a request to the device."""
+        partial_function = partial(function, *args, **kwargs)
+        for attempt in range(DEFAULT_RETRY):
```

Comment:
I think we could move this logic to the library in the future. Each packet has a counter that serves as an identifier. If the device receives two requests with the same identifier it does not execute them twice, which is good.

But we have no control over that here. If a delay occurs and the device receives more than one request for the same command, it will execute both because we are sending the packets with different counters.